File: cpp/lib/mod/nCk_mod_p.cpp
#include <bits/stdc++.h>
#define rep(i,n) for (int i = 0; i < (n); ++i)
using namespace std;
using ll = long long;


class MODCombination {
    public:
        MODCombination(int n, int p) {
            _n = n;
            _p = p;
            _make_fac_tables_mod_p();
        }

        ll nCk(int n, int k) { return ((_fac[n] * _ifac[n - k]) % _p) * _ifac[k] % _p; }

    private:
        int _n, _p;
        vector<ll> _fac, _ifac, _inv;
        void _make_fac_tables_mod_p() {
            if (_n < 1) throw "n must be more than 0.";
            _fac = vector<ll>(_n + 1, 0);
            _ifac = vector<ll>(_n + 1, 0);
            _inv = vector<ll>(_n + 1, 0);
            _fac[0] = 1; _fac[1] = 1;
            _ifac[0] = 1; _ifac[1] = 1;
            _inv[1] = 1;
            for (int i = 2; i <= _n; i++) {
                _fac[i] = _fac[i - 1] * i % _p;
                _inv[i] = _p - _inv[_p % i] * (_p / i) % _p;
                _ifac[i] = _ifac[i - 1] * _inv[i] % _p;
            }
        }
};


int main() {
    const int P = 1e+9 + 7;
    int H, W;
    cin >> H >> W;
    H--; W--;

    auto mod_comb = MODCombination(W + H, P);
    auto ans = mod_comb.nCk(W + H, H);
    cout << ans << endl;
    return 0;
}

/*
https://atcoder.jp/contests/abc034/tasks/abc034_c

Input 1
-------
123 456

Output 1
--------
210368064

*/

File: python/lib/mod/nCk_mod_p.py
P = 10**9 + 7


class MODCombination:
    def __init__(self, n, p):
        self.n = n
        self.p = p  # MOD
        self.fac, self.ifac = self.make_fac_tables_mod_p(n, p)

    @staticmethod
    def make_fac_tables_mod_p(n, p):
        assert n >= 1
        fac = [0] * (n + 1)   # _______ factorial: n!
        ifac = [0] * (n + 1)  # inverse factorial: n!^{-1}
        inv = [0] * (n + 1)   # inverse element  : n ^{-1}
        fac[0] = fac[1] = 1
        ifac[0] = ifac[1] = 1
        inv[1] = 1
        for i in range(2, n + 1):
            fac[i] = fac[i - 1] * i % p
            inv[i] = p - inv[p % i] * (p // i) % p
            ifac[i] = ifac[i - 1] * inv[i] % p
        return fac, ifac

    def nCk(self, n, k):
        # nCk = n! * k!^{-1} * (n-k)!^{-1}
        return ((self.fac[n] * self.ifac[n - k]) % self.p) * self.ifac[k] % self.p


def main():
    print("n, k:")
    n, k = map(int, input().split())

    mod_comb = MODCombination(n, P)
    ans = mod_comb.nCk(n, k)
    print(ans)


if __name__ == "__main__":
    main()
